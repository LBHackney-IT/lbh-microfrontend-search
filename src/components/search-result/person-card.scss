@import '~lbh-frontend/lbh/base';

.mtfh-search-person {
    .mtfh-search-person {
        &__title {
            @include lbh-heading-h4;
            margin: 0;

            &:link {
                color: lbh-colour(lbh-text);
                text-decoration: none;
            }

            &:visited,
            &:hover {
                color: lbh-colour(lbh-text);
            }

            &:hover {
                text-decoration: underline;
            }
        }
        &__subtitle {
            @include lbh-heading-h6;
        }
        &__row {
            @include lbh-body-xs;
            color: rgba(0, 0, 0, 0.54);

            strong {
                color: #000000;
            }
        }
    }
}
