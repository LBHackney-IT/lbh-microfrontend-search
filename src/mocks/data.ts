import faker from 'faker';
import { Tenure, HouseholdMember } from '@mtfh/common/lib/api/tenure/v1';
import {
    PersonSearchResult,
    TenureSummary,
    IdentificationTypes,
} from '@mtfh/common/lib/api/person/v1';
import { AssetSearchResult } from '../types';

faker.seed(1);

const generateMockTenureSummary = (): TenureSummary => ({
    id: faker.datatype.uuid(),
    type: 'secure',
    startDate: faker.date.past(1).toISOString().slice(0, 10),
    endDate: faker.date.future(1).toISOString().slice(0, 10),
    assetFullAddress: faker.address.streetAddress(),
    assetId: faker.datatype.uuid(),
    isActive: true,
    paymentReference: '',
    propertyReference: '',
    uprn: faker.datatype
        .number({ max: 99999999999 })
        .toString()
        .padStart(11, '0'),
});

export const generateMockPerson = (): PersonSearchResult => ({
    id: faker.datatype.uuid(),
    title: faker.name.prefix(),
    firstname: faker.name.firstName(),
    middleName: faker.name.middleName(),
    preferredFirstname: '',
    preferredSurname: '',
    surname: faker.name.lastName(),
    // Subset of ethnic groups https://www.ethnicity-facts-figures.service.gov.uk/style-guide/ethnic-groups
    ethinicity: faker.helpers.randomize([
        'English',
        'Irish',
        'Indian',
        'Chinese',
        'African',
        'Caribbean',
        'Arab',
    ]),
    // SUbset of nationalities https://www.gov.uk/government/publications/nationalities/list-of-nationalities
    nationality: faker.helpers.randomize([
        'Afghan',
        'Bahamian',
        'Cambodian',
        'Danish',
        'Egyptian',
        'French',
        'German',
        'Haitian',
        'Icelandic',
    ]),
    placeOfBirth: faker.address.cityName(),
    dateOfBirth: faker.date.past(30).toISOString().slice(0, 10),
    gender: faker.helpers.randomize(['M', 'F']),
    identification: [
        {
            identificationType: IdentificationTypes.NI,
            value: faker.random.alphaNumeric(8),
            isOriginalDocumentSeen: false,
        },
    ],
    personTypes: ['Tenant'],
    isPersonCautionaryAlert: faker.datatype.boolean(),
    isTenureCautionaryAlert: faker.datatype.boolean(),
    tenures: Array.from({ length: 3 }).map<TenureSummary>(() =>
        generateMockTenureSummary()
    ),
});

export const mockPersons = Array.from({ length: 40 }).map(() =>
    generateMockPerson()
);

const tenureTypes = {
    LEA: 'Leasehold (RTB)',
    NON: 'Non-Secure',
    SEC: 'Secure',
    TLA: 'Temp Annex',
    TBB: 'Temp B&B',
};

export const generateMockTenure = (): Tenure => {
    const isActive = faker.datatype.boolean();
    const startOfTenureDate = faker.date.past();
    const endOfTenureDate = isActive
        ? faker.datatype.boolean()
            ? faker.date.future()
            : null
        : faker.date.between(startOfTenureDate, new Date());
    const tenureType = faker.random.arrayElement(Object.keys(tenureTypes));
    return {
        id: faker.datatype.uuid(),
        paymentReference: faker.datatype
            .number({ max: 999999999 })
            .toString()
            .padStart(9, '0'),
        startOfTenureDate: startOfTenureDate.toISOString().slice(0, 10),
        endOfTenureDate:
            endOfTenureDate && endOfTenureDate.toISOString().slice(0, 10),
        householdMembers: [
            ...Array.from({ length: faker.datatype.number(4) }).map(() => ({
                id: faker.datatype.uuid(),
                fullName: faker.name.findName(),
                dateOfBirth: faker.date.past(30).toISOString().slice(0, 10),
                isResponsible: true,
                type: 'person' as HouseholdMember['type'],
            })),
            ...Array.from({ length: faker.datatype.number(4) }).map(() => ({
                id: faker.datatype.uuid(),
                fullName: faker.name.findName(),
                dateOfBirth: faker.date.past(30).toISOString().slice(0, 10),
                isResponsible: false,
                type: 'person' as HouseholdMember['type'],
            })),
        ],
        tenureType: {
            code: tenureType,
            description: tenureTypes[tenureType as keyof typeof tenureTypes],
        },
        tenuredAsset: {
            id: faker.datatype.uuid(),
            fullAddress: faker.address.streetAddress(),
            type: 'Dwelling',
            uprn: faker.datatype
                .number({ max: 99999999999 })
                .toString()
                .padStart(11, '0'),
        },
        // here
        propertyReference: '2653838280',
        accountType: { code: 'S', description: 'Service Charge' },
        charges: {
            rent: 0,
            currentBalance: 0,
            billingFrequency: '01',
            paymentReference: '299049788',
            rentGroupCode: 'LSC',
            rentGroupDescription: 'LH Serv Charges',
            serviceCharge: 0,
            otherCharges: 0,
            combinedServiceCharges: 0,
            combinedRentCharges: 0,
            tenancyInsuranceCharge: 0,
            originalRentCharge: 0,
            originalServiceCharge: 0,
        },
        isActive: false,
        isTenanted: false,
        terminated: { isTerminated: true, reasonForTermination: '' },
        successionDate: '1900-01-01T00:00:00',
        masterAccountTenureReference: '',
        evictionDate: '1900-01-01T00:00:00',
        potentialEndDate: '1900-01-01T00:00:00',
        rentCostCentre: 'H2951',
        isMutualExchange: false,
        informHousingBenefitsForChanges: false,
        isSublet: true,
        subletEndDate: '1900-01-01T00:00:00',
        subsidiaryAccountsReferences: [],
        notices: [
            {
                type: '',
                servedDate: '1900-01-01T00:00:00',
                expiryDate: '1900-01-01T00:00:00',
                effectiveDate: '1900-01-01T00:00:00',
                endDate: 'null',
            },
        ],
        legacyReferences: [
            { name: 'uh_tag_ref', value: '0905836/01' },
            { name: 'u_saff_tenancy', value: '' },
        ],
        agreementType: { code: 'M', description: 'Master Account' },
    };
};

export const mockTenures = Array.from({ length: 40 }).map(() =>
    generateMockTenure()
);

export const mockAssetSearchResult: AssetSearchResult = {
    id: '986a2a9e-9eb4-0966-120a-238689e3e265',
    assetId: '00054374',
    assetType: 'LettableNonDwelling',
    assetAddress: {
        uprn: '100021065786',
        addressLine1: 'Powell Road',
        addressLine2: 'Hackney',
        addressLine3: 'London',
        addressLine4: '',
        postCode: 'E5 8DH',
        postPreamble: '1 Newcome House',
    },
    tenure: {
        id: '0bb55bde-bc73-d7fe-0324-5cee5b59bc8c',
        paymentReference: '228008546',
        type: 'SECURE',
        startOfTenureDate: '2004-09-16',
        endOfTenureDate: '2021-08-03',
        isActive: true,
    },
};
