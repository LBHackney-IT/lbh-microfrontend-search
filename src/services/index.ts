export * from './search';
export * from './localCache';
export { config } from './config';
export { default as locale } from './locale';
