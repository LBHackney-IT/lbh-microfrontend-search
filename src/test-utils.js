import { server } from './mocks';

global.fetch = require('node-fetch');

Object.defineProperty(window, 'sessionStorage', {
    value: {
        setItem: jest.fn(),
        getItem: jest.fn(),
    },
});

beforeAll(() => {
    server.listen();
});

afterEach(() => {
    server.resetHandlers();
});

afterAll(() => {
    server.close();
});
