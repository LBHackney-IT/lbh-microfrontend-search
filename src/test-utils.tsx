import { SWRConfig } from 'swr';
import { Route, Router } from 'react-router-dom';
import React from 'react';
import { rest } from 'msw';
import MatchMediaMock from 'jest-matchmedia-mock';
import { MemoryHistory, createMemoryHistory } from 'history';
import { RenderResult, render, cleanup } from '@testing-library/react';
import { queries } from '@mtfh/common';

import { server } from './mocks';

let session: Record<string, string> = {};

Object.defineProperty(window, 'sessionStorage', {
    value: {
        setItem: jest.fn((key, value) => {
            session[key] = value;
        }),
        getItem: jest.fn(key => session[key] || null),
    },
});

let matchMedia: MatchMediaMock;

beforeAll(() => {
    matchMedia = new MatchMediaMock();
    server.listen();
});

afterEach(async () => {
    cleanup();
    server.resetHandlers();
    matchMedia.clear();
    session = {};
});

afterAll(() => {
    server.close();
});

interface RouteRenderConfig {
    url: string;
    path: string;
    query: keyof typeof queries;
}

export const routeRender = (
    component: JSX.Element,
    options?: Partial<RouteRenderConfig>
): { result: RenderResult; history: MemoryHistory } => {
    const config = {
        url: '/search',
        path: '/search',
        ...options,
    };

    const history = createMemoryHistory();
    history.push(config.url);

    const result = render(
        <SWRConfig
            value={{
                provider: () => new Map(),
                dedupingInterval: 0,
                shouldRetryOnError: false,
                errorRetryCount: 0,
            }}
        >
            <Router history={history}>
                <Route path={config.path}>{component}</Route>
            </Router>
        </SWRConfig>
    );

    return {
        result,
        history,
    };
};

export const get = (
    path: string,
    data: Record<string, any>,
    code = 200
): void => {
    server.use(
        rest.get(path, (req, res, ctx) => {
            return res(ctx.status(code), ctx.json(data));
        })
    );
};

window.HTMLElement.prototype.scrollIntoView = jest.fn();
