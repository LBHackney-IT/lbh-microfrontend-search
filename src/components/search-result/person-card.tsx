import { Link as RouterLink } from 'react-router-dom';
import React, { ComponentPropsWithoutRef } from 'react';
import { parseISO, format } from 'date-fns';
import cn from 'classnames';
import { PersonSearchResult } from '@mtfh/common/lib/api/person/v1';
import { Link, LinkBox, LinkOverlay } from '@mtfh/common';
import { SearchCard } from '../search-card';
import { locale } from '../../services';

interface PersonCardProps
    extends Omit<ComponentPropsWithoutRef<'div'>, 'children'> {
    person: PersonSearchResult;
}

const { personOriginalFullName, multipleTenures, tenureType } = locale.person;

export const PersonCard = ({
    person,
    className,
    ...props
}: PersonCardProps): JSX.Element => {
    const hasTenure = person.tenures.length > 0;
    const tenure = person.tenures[0] || {};
    const { assetFullAddress: address, type = 'N/A' } = tenure;
    const isMultipleTenancies = person.tenures.length > 1;
    return (
        <LinkBox className={cn('mtfh-search-person', className)}>
            <SearchCard {...props}>
                <LinkOverlay>
                    <Link
                        as={RouterLink}
                        to={`/person/${person.id}`}
                        variant="text-colour"
                    >
                        {personOriginalFullName(person)}
                    </Link>
                </LinkOverlay>
                {!isMultipleTenancies && address && (
                    <p>
                        <strong>{address}</strong>
                    </p>
                )}
                {isMultipleTenancies && (
                    <p>
                        <strong>{multipleTenures}</strong>
                    </p>
                )}
                <p>
                    <strong>DOB:</strong>{' '}
                    {format(parseISO(person.dateOfBirth), 'dd/MM/yyyy')}
                </p>

                {hasTenure && !isMultipleTenancies && (
                    <p>
                        <strong>{tenureType}:</strong> {type}
                    </p>
                )}
            </SearchCard>
        </LinkBox>
    );
};
