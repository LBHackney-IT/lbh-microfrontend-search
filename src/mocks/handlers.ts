import { rest } from 'msw';

import { mockPersons, mockTenures } from './data';
import { config } from '../services';

const data = {
    persons: mockPersons,
    tenures: mockTenures,
};

export const handlers = [
    rest.get(
        `${config.searchApiUrl}/search/:type`,
        (request, response, context) => {
            const { type } = request.params;
            const parameters = request.url.searchParams;
            const page = parameters.has('page')
                ? Number(parameters.get('page'))
                : 1;
            const size = parameters.has('pageSize')
                ? Number(parameters.get('pageSize'))
                : 1;
            const isDesc = parameters.has('isDesc')
                ? parameters.get('isDesc') === 'true'
                : false;
            const sortBy = parameters.has('sortBy')
                ? (parameters.get('sortBy') as string)
                : null;

            let results = data[type as keyof typeof data] as any[];
            if (sortBy) {
                results = results.sort((a, b) => {
                    const sort = sortBy as keyof typeof a;
                    if (sortBy in a) {
                        if (a[sort] > b[sort]) {
                            return isDesc ? -1 : 1;
                        } else if (a[sort] < b[sort]) {
                            return isDesc ? 1 : -1;
                        }
                    }

                    return 0;
                });
            }

            const pagedResults = results.slice((page - 1) * size, page * size);

            return response(
                context.status(200),
                context.json({
                    results: { [type]: pagedResults },
                    total: results.length,
                })
            );
        }
    ),
];
