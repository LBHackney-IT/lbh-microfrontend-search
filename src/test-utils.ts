import { server } from "@hackney/mtfh-test-utils";
import { rest } from "msw";

import { handlers } from "./mocks/handlers";

let session: Record<string, string> = {};

Object.defineProperty(window, "sessionStorage", {
  value: {
    setItem: jest.fn((key, value) => {
      session[key] = value;
    }),
    getItem: jest.fn((key) => session[key] || null),
  },
});

beforeEach(() => {
  server.use(...handlers);
  session = {};
});

export const get = (
  path: string,
  data: Record<string, any>,
  code = 200
): void => {
  server.use(
    rest.get(path, (req, res, ctx) => {
      return res(ctx.status(code), ctx.json(data));
    })
  );
};
