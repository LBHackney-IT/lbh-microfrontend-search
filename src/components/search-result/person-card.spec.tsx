import React from 'react';
import { screen } from '@testing-library/react';
import { PersonCard } from './person-card';
import { routeRender } from '../../test-utils';
import { locale } from '../../services';
import { mockPersons } from '../../mocks';

test('it renders the person card', () => {
    const {
        result: { container },
    } = routeRender(<PersonCard person={mockPersons[0]} />);
    expect(container).toMatchSnapshot();
});

test('it renders a person without tenures', () => {
    const {
        result: { container },
    } = routeRender(<PersonCard person={{ ...mockPersons[0], tenures: [] }} />);
    expect(container).toMatchSnapshot();
});

test('it renders a person with one tenure', () => {
    routeRender(
        <PersonCard
            person={{ ...mockPersons[0], tenures: [mockPersons[0].tenures[0]] }}
        />
    );
    expect(
        screen.queryByText(locale.person.multipleTenures)
    ).not.toBeInTheDocument();
});
