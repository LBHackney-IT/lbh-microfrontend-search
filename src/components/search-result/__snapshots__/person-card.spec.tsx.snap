// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a person without tenures 1`] = `
<div>
  <div
    class="mtfh-link-box mtfh-search-person"
  >
    <div
      class="mtfh-search-card"
    >
      <div
        class="mtfh-link-overlay"
      >
        <a
          class="govuk-link lbh-link lbh-link--text-colour lbh-link--no-visited-state"
          href="/person/6fbe024f-2316-4265-a6e8-d65a837e308a"
        >
          Dr. Haylee b Macejkovic
        </a>
      </div>
      <p>
        <strong>
          DOB:
        </strong>
         
        09/10/2015
      </p>
    </div>
  </div>
</div>
`;

exports[`it renders the person card 1`] = `
<div>
  <div
    class="mtfh-link-box mtfh-search-person"
  >
    <div
      class="mtfh-search-card"
    >
      <div
        class="mtfh-link-overlay"
      >
        <a
          class="govuk-link lbh-link lbh-link--text-colour lbh-link--no-visited-state"
          href="/person/6fbe024f-2316-4265-a6e8-d65a837e308a"
        >
          Dr. Haylee b Macejkovic
        </a>
      </div>
      <p>
        <strong>
          Multiple Tenures
        </strong>
      </p>
      <p>
        <strong>
          DOB:
        </strong>
         
        09/10/2015
      </p>
    </div>
  </div>
</div>
`;
